lib/LaTeXML/Core/Gullet.pm

```diff
@@ -338,47 +335,47 @@ sub readXToken {
           $self->handleMarker($token); } } }
     if (!defined $token) {
       return unless $$self{autoclose} && $toplevel && @{ $$self{mouthstack} };
-      $self->closeMouth; }                                       # Next input stream.
-    elsif (my $unexpanded = $$token[2]) {                        # Inline get_dont_expand
-      return $token; }                                           # Defer expansion (recursion?)
-                                                                 # Wow!!!!! See TeX the Program \S 309
+      $self->closeMouth; }    # Next input stream.
+        # Handle \noexpand and  smuggled tokens; either expand to $$token[2] or defer till later
+    elsif (my $unexpanded = $$token[2]) {    # Inline get_dont_expand
+      return ($cc != CC_SMUGGLE_THE) || $LaTeXML::SMUGGLE_THE ? $token : $unexpanded; }
+    ## Wow!!!!! See TeX the Program \S 309
     elsif (!$LaTeXML::ALIGN_STATE    # SHOULD count nesting of { }!!! when SCANNED (not digested)
       && $LaTeXML::READING_ALIGNMENT
       && (($atoken, $atype, $ahidden) = $self->isColumnEnd($token))) {
       $self->handleTemplate($LaTeXML::READING_ALIGNMENT, $token, $atype, $ahidden); }
-    ## Note: special-purpose lookup in State, for efficiency
-    elsif (defined($defn = LaTeXML::Core::State::lookupExpandable($STATE, $token, $toplevel))) {
+    ## Note: use general-purpose lookup, since we may reexamine $defn below
+    elsif ($LaTeXML::Core::State::CATCODE_ACTIVE_OR_CS[$$token[1]]
+      && defined($defn = LaTeXML::Core::State::lookupMeaning($STATE, $token))
+      && ((ref $defn) ne 'LaTeXML::Core::Token')    # an actual definition
+      && $$defn{isExpandable}
+      && ($toplevel || !$$defn{isProtected})) {     # is this the right logic here? don't expand unless di
       local $LaTeXML::CURRENT_TOKEN = $token;
-      my $invoked   = $defn->invoke($self) || [];
-      my @expansion = ();
-      for my $exp_t (@$invoked) {
-        my $r = ref $exp_t;
-        if ($r eq 'LaTeXML::Core::Token') {
-          push @expansion, $exp_t; }
-        elsif ($r eq 'LaTeXML::Core::Tokens') {
-          push @expansion, @$exp_t; }
-        else {
-          Fatal('misdefined', $r, undef, "Expected a Token, got " . Stringify($_)); } }
+      my $r;
+      my @expansion = map { (($r = ref $_) eq 'LaTeXML::Core::Token' ? $_
+          : ($r eq 'LaTeXML::Core::Tokens' ? @$_
+            : Fatal('misdefined', $r, undef, "Expected a Token, got " . Stringify($_),
+              "in " . ToString($defn)))) }
```

Comment:
I'll take your word that the map is faster than the for loop, I guess you never know until you benchmark... I recall rewriting the map to the for here, so my bad if I made it slower.

Reply:
Can't say that it's true in general, but seemed so in these cases.